### ISSUE TITLE
Git urls don't allow semver ranges

### ISSUE THREAD

**Issue by NickHeiner** [npm/npm#3328]
According to the documentation:

> Git URLs as Dependencies
> Git urls can be of the form:
> 
> git://github.com/user/project.git#commit-ish
> git+ssh://user@hostname:project.git#commit-ish
> git+ssh://user@hostname/project.git#commit-ish
> git+http://user@hostname/project/blah.git#commit-ish
> git+https://user@hostname/project/blah.git#commit-ish
> 
> The commit-ish can be any tag, sha, or branch which can be supplied as an argument to git checkout. The default is master.

I would like to be able use the full semver range syntax for the `commit-ish` part:

> git://github.com/user/project.git#1.0.x
> git://github.com/user/project.git#~2

I'm dependent on git urls because my company uses a number of repos on our internal github instance, and we don't have our own npm registry. (Based on digging through some old issues, it wasn't even clear that setting up our own npm registry would be possible.)

Could I submit a PR for this?

**Comment by jasonkuhrt** [comment]
@NickHeiner How's that working out with your private packages? Even today, years since `npm` started, private package hosting is only just in beta.

**Comment by rlidwka** [comment]
> Even today, years since npm started, private package hosting is only just in beta.

npm private package hosting isn't the only one hosting available out there. I'm using private packages for 2 years now, and it works just fine.

But git urls are useful for public stuff, if people don't want to use npm registry for one reason or another.

**Comment by jasonkuhrt** [comment]
> npm private package hosting isn't the only one hosting available out there. I'm using private packages for 2 years now, and it works just fine.

Link to service provider?

**Comment by rlidwka** [comment]
@jasonkuhrt , if you are looking for a npm-compatible private registry installed on your servers, look at `sinopia`. If you're looking for 3rd party service provider (SaaS), check `cnpm`.

**Comment by kumavis** [comment]
+1 this closed issue to the moon! :rocket:

**Comment by NickHeiner** [comment]
@jasonkuhrt, we are hosting our own copy of the npm registry on couchdb. It's painful to maintain at times but generally works.

**Comment by Aaike** [comment]
how come jspm is able to install git packages using a semver range ? can the same technique not be used ? i think it's just getting the tags using `git ls-remote --tags` and then do the semver match.
it would really be fantastic to be able to specify a semver range for github packages...

**Comment by ciekawy** [comment]
+1
being impressed by the richness of node packages and the well working idea of reusable modules I am convinced that this is the very case where having an option of working semvers for git urls is worth extracting the job already done for bower.

**Comment by dehru** [comment]
bower.json has no problem with this syntax 
```
"thingy": "http://thingy.lss.com/scm/cuc/emc-login.git#1.x.x",
```
I wish I could add a similar syntax to package.json.

**Comment by ciekawy** [comment]
@dehru
Actually latest npm resolves your case correctly. The challenge is the
semver handling ie. git://myrepo.com/mymodule.git#^1.0.5

I mean the caret and tilde chars as range selectors.

**Comment by SakerONE** [comment]
@ciekawy, see https://github.com/npm/node-semver#x-ranges-12x-1x-12-

**Comment by gabrielhpugliese** [comment]
+1 to this, I'm suffering without it :(((

**Comment by othiym23** [comment]
GOOD NEWS, everybody! It may not work exactly like it does in Bower, but adding this functionality is now on npm's roadmap for the near- to medium-term, as part of making GitHub-hosted repositories work more like registry-hosted packages.

**Comment by gabrielhpugliese** [comment]
:rocket::rocket::rocket:

**Comment by FGRibreau** [comment]
@othiym23 :+1: !!!!!!!!!

**Comment by aaronjensen** [comment]
It may be more related to https://github.com/npm/npm/issues/6236, but when this is addressed, it seems like `npm update` should always reset to the specified commit-ish even if it's not a semver range. Ideally `npm install` would as well (it's still confusing that those are two different things instead of one idempotent operation)

**Comment by joshrtay** [comment]
I'd really love to see this added. Would make git installs way more robust. What's the timeline?

**Comment by gasteve** [comment]
This is badly needed.

**Comment by d3mi3n** [comment]
is this still part of the roadmap?

**Comment by KenanY** [comment]
@d3mi3n [yes](https://github.com/npm/npm/wiki/Roadmap-area-of-focus:-git)

**Comment by othiym23** [comment]
Yes, with the possibility that we may not get to it this year. We're in the early stages of pretty drastically trimming down the CLI's roadmap to match what we reasonably believe we can deliver in a 6-12 month timeframe. Expect to hear more about this from the team soon.

**Comment by gillesdemey** [comment]
+1 for supporting semver ranges :+1:

**Comment by juanstiza** [comment]
+1 semver yay!

**Comment by raveclassic** [comment]
+1 for semver ranges!
Currently I need to maintain several proprietary projects and libraries in local bitbucket, and the only way to reference them with support of semver (**it is crucial**, because it's an overkill to raise version in each project on library patch) is using bower which is very painful to support along with npm.
Guys, this is needed realy badly...

**Comment by filipesilva** [comment]
Glad to hear this is happening. My usecase is similar to @raveclassic's one.

**Comment by florianb** [comment]
:+1:

**Comment by dimfeld** [comment]
@raveclassic @filipesilva I'm achieving this by maintaining multiple tags in git for each revision. For example:

```
git commit package.json -m "v1.5.3"
git tag v1.5.3
# Update existing major/minor version tags to point to the new commit.
git tag --force v1.5
git tag --force v1
git push --tags --force
```

Then in the other project's package.json, I get semver-ish functionality by just referencing the `v1` or `v1.5` tag:

```
"dependencies": {
    "mypackage": "git+ssh://server/mypackage.git#v1",
}
```

First, the obvious problem: changing things already pushed to a git repo is inadvisable. My team finds it to work ok here since it's only overwriting tags instead of full commits, but your tolerance of such a practice may vary. 

More importantly, if the package is already installed in a project as described above and you want to bump the version, `npm install` on its own will not pick up the changed tag, as npm reasonably assumes that established things in the repository aren't changing out from under it. `npm install --force mypackage` will refetch the repository and pick up the new version though.

So, this isn't perfect, but we find it works reasonably well in the absence of full support from npm. Hope it helps.

**Comment by raveclassic** [comment]
@dimfeld Thanks for the tip, but unfortunately this won't help with version ranging with `^` and `~` which is critical

**Comment by filipesilva** [comment]
@dimfeld thanks for pitching in your strategy! The team I'm in is a small one and as is atm, I think moving the tags would be more complicated (and thus error prone) for everyone to grok correctly than just being thorough with updating platforms.

Experience might show me otherwise, and in that case your strategy is probably the soundest one. The only other sane alternative is branches for minor version ranges, as someone suggested earlier.

**Comment by dimfeld** [comment]
Yeah, I agree it's a high-maintenance strategy... It could maybe be automated using a post-commit hook that pulls the package version from `package.json` and updates the tags if needed. https://coderwall.com/p/mk18zq/automatic-git-version-tagging-for-npm-modules may be a good starting point, though I haven't tried it yet.

As far as ^ and ~ compatibility, my thinking was that using the `vX` tag is similar to using `^X.Y.Z`, in that it get the latest version with a matching major version , and the `vX.Y` tag is like using `~X.Y.Z`, as it gets the latest version matching the major and minor versions. When you update a library from `1.5.7` to `1.6.0`, for example, the `v1.5.7` and `v1.5` tags would not be updated, so puling in the library at tag `v1.5` would continue to be similar to semver matching on `~1.5`. Of course, this is a simplified model compared to the full power of semver matching, and there may be some subtleties you're thinking of that I'm forgetting. And if you're unable to force push it's a moot point anyway.

Ultimately, I agree fully that having official support for semver ranges on git urls would be much better :)

**Comment by StephanBijzitter** [comment]
Honestly, although I highly support npm adding this possibility, why doesn't GitHub itself do this?
https://github.com/npm/npm/releases/tag/v3.8.x would be rather awesome. The only caveat would be that a tag could literally be `v3.8.x` and in that case, you'd have a pretty poor naming convention and you just don't deserve any better.

**Comment by arikon** [comment]
@StephanBijzitter This should be done in npm, not in Github.

**Comment by othiym23** [comment]
The CLI team still intends to ship this as part of giving hosted Git dependencies as much parity with registry dependencies as is practicable. If somebody wants to take a stab at implementing this, it's mostly in the caching logic (and `fetchPackageMetadata`) where things would need to change. Thanks to all for your patience!

**Comment by cellvia** [comment]
thank you @dimfeld  i ended up implementing the same thing for my team as well, only difference is i opted to use v1.1.x and v1.x.x  notation to help make obvious to them whats going on with the semver-ing ... works like a charm til semver on git urls is officially supported

**Comment by dimfeld** [comment]
I'm glad it's working well for you @cellvia. Good idea with the v1.x.x style notation. I think I'll start using that too.

**Comment by sankethkatta** [comment]
@othiym23 I took a stab at implementing this over at https://github.com/npm/npm/pull/15308. Would love to get your feedback!

**Comment by ondratra** [comment]
If you really need this feature you can just switch to *Yarn* utility instead of *NPM*. It supports regular package.json and has this feature already.

Example
```
$ yarn add visionmedia/commander.js#^1.2.0
yarn add v0.20.3
info No lockfile found.
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 2 new dependencies.
├─ commander@1.3.2
└─ keypress@0.1.0
warning No license field
Done in 1.98s.
```

**Comment by raveclassic** [comment]
@ondratra That's exactly what we've ended up with.

**Comment by benwiley4000** [comment]
Worth noting this is supported now: https://github.com/npm/npm/releases/tag/v5.0.0

**Comment by SkeLLLa** [comment]
It seems broken in npm 5 (at least 5.5.1):
```
"npm": "git+https://github.com/npm/npm#semver:^5.5.0"
```
Produces:
```
npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for npm
```

**Comment by Yannik** [comment]
@SkeLLLa maybe try `#semver:^v5.5.0`?

**Comment by Yannik** [comment]
@SkeLLLa also, according to https://github.com/npm/npm/issues/17621, this should be fixed with 5.6.0

**Comment by SkeLLLa** [comment]
@Yannik yep, it supported actually in >=5.6.0.
